**"Predicting House Sale Prices: A Comprehensive Analysis and Model Comparison"** 

This project involves predicting house sale prices using machine learning models, along with a thorough analysis and comparison of different models.

## Introduction
In this project, we aim to predict house prices using machine learning techniques. We will analyze the dataset, perform data preprocessing, feature engineering, model training, hyperparameter tuning to achieve the best predictive performance.

Let's start by importing the necessary libraries and loading the dataset.

In [7]:
#import relevant packages 
import pandas as pd 
import numpy as np 
import plotly.graph_objects as go
import plotly.express as px
import scipy.stats as stats
from IPython.display import display, HTML

In [8]:
# Provide the path to the CSV file
file_path = r'C:\Users\nasrin\Desktop\DSproject\house-prices-advanced-regression-techniques\train.csv'

# Load the dataset into a DataFrame
df = pd.read_csv(file_path)

Now creating a function `create_scrollable_table` to display a scrollable table within a small window. This function will be useful for displaying large datasets without taking up too much space on your notebook.

In [9]:
# Function to create scrollable table within a small window
def create_scrollable_table(df, table_id, title):
    html = f'<h3>{title}</h3>'
    html += f'<div id="{table_id}" style="height:200px; overflow:auto;">'
    html += df.to_html()
    html += '</div>'
    return html

## Exploratory Data Analysis (EDA)
First, let's explore the dataset to understand its structure, features, and distributions.

Exploratory Data Analysis (EDA) is an approach to analyzing datasets to summarize their main characteristics, often employing graphical representations and statistical techniques. The primary goal of EDA is to gain insights and understanding of the data, rather than making formal statistical inferences. Key aspects of EDA include:

1. **Data Summarization**: Understanding the basic properties of the dataset, such as its size, structure, and key statistical measures like mean, median, and standard deviation.

2. **Visualization**: Creating visual representations of the data using plots, histograms, scatter plots, box plots, etc., to identify patterns, trends, outliers, and relationships between variables.

3. **Data Cleaning**: Identifying and handling missing values, outliers, and errors in the data to ensure its quality and reliability for further analysis.

4. **Feature Engineering**: Creating new features or transforming existing ones to better represent the underlying patterns and relationships in the data.

5. **Dimensionality Reduction**: Techniques like principal component analysis (PCA) or t-distributed stochastic neighbor embedding (t-SNE) are used to reduce the dimensionality of the data while preserving important information.

6. **Statistical Testing**: Conducting statistical tests to validate hypotheses or assumptions about the data, such as testing for normality, independence, or correlation between variables.

Overall, EDA is a crucial step in the data analysis process as it helps analysts understand the nature of the data, uncover patterns, and generate hypotheses for further investigation.

In [53]:
# Display the shape of the dataframe
display(df.shape)

# Display the first five rows of the dataset
display(df.head())

# Display the summary statistics of the dataset
display(df.describe())

(1460, 82)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolQC Fence MiscFeature MiscVal MoSold YrSold  \
0         Lvl    AllPub  ...    NaN   NaN         NaN       0      2   2008   
1         Lvl    AllPub  ...    NaN   NaN         NaN       0      5   2007   
2         Lvl    AllPub  ...    NaN   NaN         NaN       0      9   2008   
3         Lvl    AllPub  ...    NaN   NaN         NaN       0      2   2006   
4         Lvl    AllPub  ...    NaN   NaN         NaN       0     12   2008   

  SaleType  SaleCondition  SalePrice  PropertyAge  
0       WD         Normal     208500            5  
1       WD         Normal     181500           31  
2       WD         Normal     223500            7  
3       WD        Abnorml     140000           91  
4       WD         Normal     250000            8  

[5 rows x 82 columns]

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

       OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch     PoolArea  \
count  1460.000000    1460.000000  1460.000000  1460.000000  1460.000000   
mean     46.660274      21.954110     3.409589    15.060959     2.758904   
std      66.256028      61.119149    29.317331    55.757415    40.177307   
min       0.000000       0.000000     0.000000     0.000000     0.000000   
25%       0.000000       0.000000     0.000000     0.000000     0.000000   
50%      25.000000       0.000000     0.000000     0.000000     0.000000   
75%      68.000000       0.000000     0.000000     0.000000     0.000000   
max     547.000000     552.000000   508.000000   480.000000   738.000000   

            MiscVal       MoSold       YrSold      SalePrice  PropertyAge  
count   1460.000000  1460.000000  1460.000000    1460.000000  1460.000000  
mean      43.489041     6.321918  2007.815753  180921.195890    36.547945  
std      496.123024     2.703626     1.328095   79442.502883    30.250152  
min        0.000000     1.000000  2006.000000   34900.000000     0.000000  
25%        0.000000     5.000000  2007.000000  129975.000000     8.000000  
50%        0.000000     6.000000  2008.000000  163000.000000    35.000000  
75%        0.000000     8.000000  2009.000000  214000.000000    54.000000  
max    15500.000000    12.000000  2010.000000  755000.000000   136.000000  

[8 rows x 39 columns]

The following code snippet provides summary statistics for both numerical and categorical features in the dataset.

For numerical features:
- It selects columns of numerical data types.
- Computes summary statistics including count, mean, standard deviation, minimum, 25th percentile, median (50th percentile), 75th percentile, and maximum.
- Generates a scrollable HTML table to display the summary statistics.

For categorical features:
- It selects columns of object data types (typically representing categorical variables).
- Computes summary statistics including count, unique categories, top category, and frequency of the top category.
- Generates a scrollable HTML table to display the summary statistics.


In [13]:
# Summary statistics for numerical features
numerical_features = df.select_dtypes(include=[np.number])
summary_stats = numerical_features.describe().T
html_numerical = create_scrollable_table(summary_stats, 'numerical_features', 'Summary statistics for numerical features')

display(HTML(html_numerical))

# Summary statistics for categorical features
categorical_features = df.select_dtypes(include=[object])
cat_summary_stats = categorical_features.describe().T
html_categorical = create_scrollable_table(cat_summary_stats, 'categorical_features', 'Summary statistics for categorical features')

display(HTML(html_categorical ))

The code snippet below performs several operations related to handling missing values in the dataset:

1. It calculates the number of null values for each feature in the dataset and displays the results in a scrollable HTML table.

2. It computes the percentage of missing values for each feature and presents the information in another scrollable HTML table.

3. The code explores rows with missing values by filtering the dataset to include only rows containing at least one missing value. It then displays the first few rows of this subset in a scrollable HTML table.

Finally, the code returns the column labels of the DataFrame.

In [14]:
# Null values in the dataset
null_values = df.isnull().sum()
html_null_values = create_scrollable_table(null_values.to_frame(), 'null_values', 'Null values in the dataset')

# Percentage of missing values for each feature
missing_percentage = (df.isnull().sum() / len(df)) * 100
html_missing_percentage = create_scrollable_table(missing_percentage.to_frame(), 'missing_percentage', 'Percentage of missing values for each feature')

display(HTML(html_null_values + html_missing_percentage))

# Exploring rows with missing values
rows_with_missing_values = df[df.isnull().any(axis=1)]
html_rows_with_missing_values = create_scrollable_table(rows_with_missing_values.head(), 'rows_with_missing_values', 'Rows with missing values')

display(HTML(html_rows_with_missing_values))

# Return column lavels of the Dataframe
df.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

### Explore the Dependent Variable

Before proceeding with modeling, it's essential to understand the distribution of the dependent variable, 'SalePrice'. This step involves visualizing the distribution of 'SalePrice' and checking if it follows a normal distribution. 

#### Normalize Dependent Variable

The code snippet performs the following tasks:
1. Fits a normal distribution to the 'SalePrice' data using the `stats.norm.fit()` function from the `scipy.stats` module.
2. Creates a histogram of the 'SalePrice' column to visualize its distribution, along with a fitted normal distribution overlay.
3. Generates a Q-Q plot (Quantile-Quantile plot) to compare the distribution of 'SalePrice' against a theoretical normal distribution. The plot includes a line of best fit to assess the deviation from normality.

Both visualizations help in assessing whether 'SalePrice' follows a normal distribution, which is important for certain modeling assumptions.

Note:The portion of this code is credited to [this notebook](https://www.kaggle.com/code/serigne/stacked-regressions-top-4-on-leaderboard).


In [18]:
import scipy.stats as stats
import plotly.graph_objects as go
import plotly.express as px

# Fit a normal distribution to the SalePrice data
mu, sigma = stats.norm.fit(df['SalePrice'])

# Create a histogram of the SalePrice column
hist_data = go.Histogram(x=df['SalePrice'], nbinsx=50, name="Histogram", opacity=0.75, histnorm='probability density', marker=dict(color='purple'))

# Calculate the normal distribution based on the fitted parameters
x_norm = np.linspace(df['SalePrice'].min(), df['SalePrice'].max(), 100)
y_norm = stats.norm.pdf(x_norm, mu, sigma)

# Create the normal distribution overlay
norm_data = go.Scatter(x=x_norm, y=y_norm, mode="lines", name=f"Normal dist. (μ={mu:.2f}, σ={sigma:.2f})", line=dict(color="green"))

# Combine the histogram and the overlay
fig = go.Figure(data=[hist_data, norm_data])

# Set the layout for the plot
fig.update_layout(
    title="SalePrice Distribution",
    xaxis_title="SalePrice",
    yaxis_title="Density",
    legend_title_text="Fitted Normal Distribution",
    plot_bgcolor='rgba(32, 32, 32, 1)',
    paper_bgcolor='rgba(32, 32, 32, 1)',
    font=dict(color='white')
)

# Show the plot
fig.show()

# Create a Q-Q plot
qq_data = stats.probplot(df['SalePrice'], dist="norm")
qq_fig = px.scatter(x=qq_data[0][0], y=qq_data[0][1], labels={'x': 'Theoretical Quantiles', 'y': 'Ordered Values'}, color_discrete_sequence=["purple"])
qq_fig.update_layout(
    title="Q-Q plot",
    plot_bgcolor='rgba(32, 32, 32, 1)',
    paper_bgcolor='rgba(32, 32, 32, 1)',
    font=dict(color='white')
)

# Calculate the line of best fit
slope, intercept, r_value, p_value, std_err = stats.linregress(qq_data[0][0], qq_data[0][1])
line_x = np.array(qq_data[0][0])
line_y = intercept + slope * line_x

# Add the line of best fit to the Q-Q plot
line_data = go.Scatter(x=line_x, y=line_y, mode="lines", name="Normal Line", line=dict(color="green"))

# Update the Q-Q plot with the normal line
qq_fig.add_trace(line_data)

# Show the Q-Q plot
qq_fig.show()

# Create a Q-Q plot with log transformation
qq_data_log = stats.probplot(df['SalePrice'], dist="norm", plot=None)
ordered_values_log = np.log(qq_data_log[0][1])
qq_fig_log = px.scatter(x=qq_data_log[0][0], y=ordered_values_log, labels={'x': 'Theoretical Quantiles', 'y': 'Ordered Values (log)'}, color_discrete_sequence=["purple"])
qq_fig_log.update_layout(
    title="Q-Q plot with Log Transformation",
    plot_bgcolor='rgba(32, 32, 32, 1)',
    paper_bgcolor='rgba(32, 32, 32, 1)',
    font=dict(color='white')
)

# Calculate the line of best fit for log-transformed data
slope_log, intercept_log, r_value_log, p_value_log, std_err_log = stats.linregress(qq_data_log[0][0], ordered_values_log)
line_x_log = np.array(qq_data_log[0][0])
line_y_log = intercept_log + slope_log * line_x_log

# Add the line of best fit to the Q-Q plot with log transformation
line_data_log = go.Scatter(x=line_x_log, y=line_y_log, mode="lines", name="Normal Line (Log)", line=dict(color="green"))
qq_fig_log.add_trace(line_data_log)

# Show the Q-Q plot with log transformation
qq_fig_log.show()



1. **Analyzing the Impact of Zoning Classifications on Sale Prices**: By examining the average sale prices across different zoning classifications, you can gain insights into how zoning regulations influence property values. This analysis helps identify which zoning categories tend to command higher sale prices and provides valuable information for buyers, sellers, and real estate investors.

2. **Calculating Property Age and Correlation with Sale Price**: Understanding the relationship between property age and sale price helps uncover trends in the real estate market. By calculating the correlation between these variables, you can determine if older or newer properties tend to sell for higher prices. This information is crucial for buyers assessing the value of a property and sellers setting asking prices.

3. **Sale Price Trends Over the Years**: Analyzing sale price trends over multiple years provides valuable insights into the dynamics of the real estate market. By visualizing how sale prices fluctuate over time, you can identify patterns, cycles, or trends that may influence buying and selling decisions. This analysis helps stakeholders make informed decisions about when to buy, sell, or invest in real estate.

Each of these analyses contributes to a comprehensive understanding of the factors influencing property values and market trends, empowering stakeholders to make informed decisions in the real estate industry.

In [30]:
import pandas as pd
import plotly.express as px

# Provide the correct path to the CSV file
file_path = r'C:\Users\nasrin\Desktop\DSproject\house-prices-advanced-regression-techniques\train.csv'

# Load the dataset into a DataFrame
df = pd.read_csv(file_path)

In [43]:
# i. Analyze the impact of zoning classifications on sale prices
zoning_prices = df.groupby('MSZoning')['SalePrice'].mean()
fig = px.bar(x=zoning_prices.index, y=zoning_prices.values, title='Average Sale Price by Zoning',
             color_discrete_sequence=['purple', 'yellow'], text=zoning_prices.values,
             labels={'x': 'Zoning Classification', 'y': 'Average Sale Price'}, template='plotly_dark')

fig.update_traces(texttemplate='$%{text:,.0f}', textposition='outside')
fig.update_yaxes(tickprefix='$', tickformat=',')
fig.show()

The bar chart illustrates the average sale price of properties categorized by zoning classification. It provides insights into how different zoning classifications impact property values. 
- Notably, the Floating Village Residential (FV) zoning classification demonstrates the highest average sale price at $214,014, indicating potentially higher property values in such areas. 
- On the other hand, Residential Medium Density (RM) zoning has the lowest average sale price at $126,317. 

This visualization helps identify trends and disparities in property values based on zoning classifications, which can be valuable for urban planning, real estate investment decisions, and market analysis.

In [44]:
# ii. Calculate Property Age
df['PropertyAge'] = df['YrSold'] - df['YearBuilt']

# Calculate Correlation between Property Age and Sale Price
age_price_corr = df['PropertyAge'].corr(df['SalePrice'])
print(f'Correlation between Property Age and Sale Price: {age_price_corr}')

# Create a scatter plot to visualize the relationship between Property Age and Sale Price
fig9 = px.scatter(df, x='PropertyAge', y='SalePrice', title='Property Age vs Sale Price', color='PropertyAge', color_continuous_scale=px.colors.sequential.Purp)

fig9.update_layout(plot_bgcolor='rgb(30,30,30)', paper_bgcolor='rgb(30,30,30)', font=dict(color='white'))

fig9.show()

Correlation between Property Age and Sale Price: -0.523350417546816


As for the output, it indicates the correlation coefficient between the property age and sale price, which is approximately -0.52 and the plots are going towards downwards also. 

This correlation value along with the graph suggests a moderate negative correlation between property age and sale price, meaning that as the age of the property increases, the sale price tends to decrease.

The following code will generate box plot. By visualizing the box plot we can describe the distribution of sale prices over the years from 2006 to 2010. The box plot provides a summary of the central tendency, dispersion, and skewness of the sale prices for each year. 

1. **Interpretation of the Box Plot**:
   - The box represents the interquartile range (IQR), with the lower and upper boundaries indicating the 25th and 75th percentiles, respectively.
   - The line within the box represents the median (50th percentile) sale price for each year.
   - The whiskers extend to the minimum and maximum sale prices within 1.5 times the IQR from the lower and upper quartiles, respectively.
   - Any points beyond the whiskers are considered outliers.

2. **Description of the Trend**:
   - From the box plot, we observe the median sale price for each year along with the spread of sale prices.
   - The annotations on the plot provide the average sale price for each year, offering additional insight into the central tendency of the data.
   - It's important to note that these sale prices represent the average sale price for properties sold in each respective year.

3. **Comparison Over Time**:
   - By examining the box plot, we can compare the distribution of sale prices across different years.
   - We can observe any trends or fluctuations in sale prices over the five-year period.
   - The annotations provide a quick reference to compare the average sale prices for each year.

4. **Addressing the Question of Property Consistency**:
   - The sale prices depicted in the box plot represent the distribution of sale prices for properties sold in each respective year.
   - While the same property could be sold in different years, the plot illustrates the overall distribution of sale prices across all properties sold in each year, rather than tracking the sale price of a specific property over time.

Overall, the box plot effectively summarizes the variation in sale prices over the five-year period, allowing for comparisons between different years and providing insights into the central tendency and spread of sale prices.

In [45]:
# iii. Box plot of price over the years
yearly_avg_sale_price = df.groupby('YrSold')['SalePrice'].mean()

fig13 = px.box(df, x='YrSold', y='SalePrice', title='Sale Price Trends Over the Years',
               points=False, color_discrete_sequence=['green'])

fig13.add_trace(px.line(x=yearly_avg_sale_price.index, y=yearly_avg_sale_price.values).data[0])

fig13.update_traces(line=dict(color='purple', width=4), selector=dict(type='scatter', mode='lines'))

for year, avg_price in yearly_avg_sale_price.items():
    fig13.add_annotation(
        x=year,
        y=avg_price,
        text=f"{avg_price:,.0f}",
        font=dict(color='white'),
        showarrow=False,
        bgcolor='rgba(128, 0, 128, 0.6)'
    )

fig13.update_layout(
    plot_bgcolor='rgb(30,30,30)',
    paper_bgcolor='rgb(30,30,30)',
    font=dict(color='white'),
    xaxis_title='Year Sold',
    yaxis_title='Sale Price'
)

fig13.show()

Overall, the box plot provides a visual summary of the distribution of sale prices for properties sold in each year from 2006 to 2010. 
### Year 2006:
- **Maximum Sale Price**: $625,000
- **Upper Quartile (Q3)**: $219,210
- **Median Sale Price**: $163,995
- **Lower Quartile (Q1)**: $131,000
- **Minimum Sale Price**: $35,311

### Year 2007:
- **Maximum Sale Price**: $755,000
- **Upper Quartile (Q3)**: $219,500
- **Median Sale Price**: $167,000
- **Lower Quartile (Q1)**: $129,800
- **Minimum Sale Price**: $39,311

### Year 2008:
- **Maximum Sale Price**: $446,261
- **Upper Quartile (Q3)**: $207,000
- **Median Sale Price**: $164,000
- **Lower Quartile (Q1)**: $131,000
- **Minimum Sale Price**: $40,000

### Year 2009:
- **Maximum Sale Price**: $582,933
- **Upper Quartile (Q3)**: $213,000
- **Median Sale Price**: $162,000
- **Lower Quartile (Q1)**: $125,000
- **Minimum Sale Price**: $34,900

### Year 2010:
- **Maximum Sale Price**: $611,657
- **Upper Quartile (Q3)**: $213,375
- **Median Sale Price**: $155,000
- **Lower Quartile (Q1)**: $128,050
- **Minimum Sale Price**: $55,000

These statistics provide a detailed breakdown of the distribution of sale prices for each year, including measures of central tendency (median) and spread (quartiles). It helps in understanding the variability and range of sale prices across different years, allowing for comparisons and insights into market trends over time.

## Model Building and Evaluation
This phase follows the data preprocessing and feature engineering steps and involves training machine learning models on the prepared dataset, tuning their hyperparameters, and evaluating their performance.

Here's how these snippets fit into the data science project phases:

1. **Data Preprocessing and Feature Engineering**: This phase involves cleaning the data, handling missing values, encoding categorical variables, and scaling numerical features. It prepares the dataset for model training. The code snippets provided earlier, such as creating a data pipeline and performing PCA (Principal Component Analysis), are part of this phase.

2. **Model Building and Evaluation**: This phase focuses on selecting appropriate machine learning models, training them on the preprocessed data, and evaluating their performance. The code snippets you provided train and tune several regression models (Linear Regression, Random Forest, XGBoost) using cross-validation and grid search for hyperparameter tuning. They also evaluate the models' performance metrics such as Root Mean Squared Error (RMSE) and Mean Absolute Error (MAE) on a holdout test set.

3. **Model Deployment and Monitoring**: After selecting the best-performing model, it can be deployed into production to make predictions on new data. Monitoring the model's performance over time and retraining it periodically with new data are also crucial aspects of this phase.

So, the code snippets are specifically executed in the "Model Building and Evaluation" phase, where the focus is on training and evaluating machine learning models to find the best-performing one for the given task.

## Data Preprocessing
The following code is for creating a data pipeline. Data pipelines are a crucial component of the data preprocessing step in a data science project. They help streamline and automate the process of transforming raw data into a format suitable for machine learning models. 

In this specific code, the pipeline is constructed using scikit-learn's Pipeline and ColumnTransformer classes. The pipeline consists of several transformation steps, including imputation for missing values, scaling numerical features, and one-hot encoding categorical features. This step ensures that the data preprocessing is consistent across the training and testing datasets, allowing for easier model training and evaluation. 

This code snippet would typically be part of the data preprocessing phase, which is an essential step in any data science project.


In [46]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder

# Define transformers for numerical and categorical columns
numerical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

In [47]:

# Update categorical and numerical columns
categorical_columns = df.select_dtypes(include=['object', 'category']).columns
numerical_columns = df.select_dtypes(include=['int64', 'float64']).columns

# Remove target variable from numerical columns
numerical_columns = numerical_columns.drop('SalePrice')

# Combine transformers using ColumnTransformer
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_columns),
        ('cat', categorical_transformer, categorical_columns)
    ],remainder = 'passthrough')

# Create a pipeline with the preprocessor
pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor)])

# Apply the pipeline to your dataset
X = df.drop('SalePrice', axis=1)
y = np.log(df['SalePrice']) #normalize dependent variable 
X_preprocessed = pipeline.fit_transform(X)

The following code snippet is performing Principal Component Analysis (PCA), which is a dimensionality reduction technique commonly used in data science projects. Here's what each part of the code does:

1. **Instantiate PCA**: PCA is instantiated with the `svd_solver` parameter set to 'arpack'. This solver is suitable for sparse matrices.

2. **Fit and transform the data**: The PCA model is fitted to the preprocessed data (`X_preprocessed`) using the `fit_transform` method. This step calculates the principal components and transforms the data into the reduced dimensional space.

3. **Calculate cumulative explained variance**: The cumulative explained variance of the principal components is calculated using the `explained_variance_ratio_` attribute of the PCA model. This provides insight into how much variance in the original data is preserved by each principal component.

4. **Choose the number of components**: The number of principal components is determined based on a threshold of explained variance. In this case, the threshold is set to 95%, and the number of components is chosen as the index where the cumulative explained variance first exceeds or equals this threshold.

5. **Instantiate PCA with the chosen number of components**: PCA is instantiated again, this time specifying the number of components determined in the previous step.

6. **Create a pipeline**: A pipeline is created that combines the preprocessing steps (such as imputation and one-hot encoding) with PCA. This ensures that the same preprocessing steps are applied to both the training and testing datasets.

7. **Fit and transform the data with PCA**: The pipeline is used to fit and transform the original data (`X`) using the PCA model with the chosen number of components. This results in the data being transformed into the reduced dimensional space.

Overall, this code snippet is an essential part of dimensionality reduction and preprocessing in a data science project, helping to reduce the complexity of the data while preserving as much relevant information as possible.

In [51]:
from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline

# Instantiate PCA with the "arpack" solver
pca = PCA(svd_solver='arpack')

# Fit and transform the data
X_pca_pre = pca.fit_transform(X_preprocessed)

# Calculate the cumulative explained variance
cumulative_explained_variance = np.cumsum(pca.explained_variance_ratio_)

# Choose the number of components based on the explained variance threshold
n_components = np.argmax(cumulative_explained_variance >= 0.95) + 1

pca = PCA(n_components=n_components, svd_solver='arpack')  # Specify solver here
pipeline_pca = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('pca', pca)
])

X_pca = pipeline_pca.fit_transform(X)

## Model Training and Hyperparameter Tuning
This code block corresponds to the **Model Training and Hyperparameter Tuning** step in the data science project. In this step, various machine learning models are trained on the training dataset, and their hyperparameters are tuned using techniques like grid search and cross-validation. The goal is to find the best-performing model configuration that minimizes the chosen evaluation metric (in this case, negative mean squared error) on the training data.

By training and tuning multiple models, data scientists aim to identify the most suitable algorithm and hyperparameters for the given dataset, which can ultimately lead to better model performance and generalization to unseen data. This step is crucial for building robust and accurate predictive models in data science projects.

This code block trains and tunes three different regression models: Linear Regression, Random Forest, and XGBoost. Here's a breakdown of what each part of the code does:

1. **Importing Libraries**: The code imports necessary libraries, including regression models (Linear Regression, Random Forest, XGBoost), grid search for hyperparameter tuning (GridSearchCV), cross-validation (KFold), and data splitting (train_test_split).

2. **Splitting Data**: The dataset is split into training and testing sets using a 80-20 split ratio.

3. **Defining Models**: Three regression models (Linear Regression, Random Forest, XGBoost) are instantiated with default hyperparameters.

4. **Defining Hyperparameter Grids**: Hyperparameter grids are defined for each model. These grids specify the hyperparameters and their respective values that will be searched during the grid search process.

5. **Cross-Validation Setup**: A 3-fold cross-validation strategy is set up using KFold, with shuffling enabled.

6. **Training and Tuning Models**: For each model, a GridSearchCV object is created with the model, hyperparameter grid, cross-validation strategy, and scoring metric (negative mean squared error). The `fit` method of the GridSearchCV object is called to perform grid search and hyperparameter tuning on the training data. The best parameters and corresponding root mean squared error (RMSE) are printed for each model.

This code block is essential for training and optimizing regression models, ensuring that the models are fine-tuned to perform well on the given dataset.

In [50]:
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor
from sklearn.model_selection import GridSearchCV, KFold, train_test_split
from sklearn.metrics import mean_squared_error

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_preprocessed, y, test_size=0.2, random_state=42)

# Define the models
models = {
    'LinearRegression': LinearRegression(),
    'RandomForest': RandomForestRegressor(random_state=42),
    'XGBoost': XGBRegressor(random_state=42)
}

# Define the hyperparameter grids for each model
param_grids = {
    'LinearRegression': {},
    'RandomForest': {
        'n_estimators': [100, 200, 500],
        'max_depth': [None, 10, 30],
        'min_samples_split': [2, 5, 10],
    },
    'XGBoost': {
        'n_estimators': [100, 200, 500],
        'learning_rate': [0.01, 0.1, 0.3],
        'max_depth': [3, 6, 10],
    }
}

# 3-fold cross-validation
cv = KFold(n_splits=3, shuffle=True, random_state=42)

# Train and tune the models
grids = {}
for model_name, model in models.items():
    grids[model_name] = GridSearchCV(estimator=model, param_grid=param_grids[model_name], cv=cv, scoring='neg_mean_squared_error', n_jobs=-1, verbose=2)
    grids[model_name].fit(X_train, y_train)
    best_params = grids[model_name].best_params_
    best_score = np.sqrt(-1 * grids[model_name].best_score_)
    
    print(f'Best parameters for {model_name}: {best_params}')
    print(f'Best RMSE for {model_name}: {best_score}\n')

# Evaluate model performance on test set
for model_name, grid_search in grids.items():
    model = grid_search.best_estimator_
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    print(f'Model: {model_name}')
    print(f'Mean Squared Error (MSE): {mse}')
    print(f'Mean Absolute Error (MAE): {np.mean(np.abs(y_test - y_pred))}')
    print(f'R-squared: {model.score(X_test, y_test)}\n')

Fitting 3 folds for each of 1 candidates, totalling 3 fits
Best parameters for LinearRegression: {}
Best RMSE for LinearRegression: 0.1761766556858557

Fitting 3 folds for each of 27 candidates, totalling 81 fits
Best parameters for RandomForest: {'max_depth': None, 'min_samples_split': 2, 'n_estimators': 500}
Best RMSE for RandomForest: 0.15328948266043066

Fitting 3 folds for each of 27 candidates, totalling 81 fits
Best parameters for XGBoost: {'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 500}
Best RMSE for XGBoost: 0.13768364339658068

Model: LinearRegression
Mean Squared Error (MSE): 0.017456402763288983
Mean Absolute Error (MAE): 0.09035825453448648
R-squared: 0.9064569226967785

Model: RandomForest
Mean Squared Error (MSE): 0.0216806238166419
Mean Absolute Error (MAE): 0.09860414458809556
R-squared: 0.8838207219916316

Model: XGBoost
Mean Squared Error (MSE): 0.01771979285489721
Mean Absolute Error (MAE): 0.08833842126303457
R-squared: 0.9050455024841315



| Model            | Mean Squared Error (MSE) | Mean Absolute Error (MAE) | R-squared |
|------------------|---------------------------|----------------------------|-----------|
| LinearRegression| 0.017456402763288983     | 0.09035825453448648      | 0.9064569226967785 |
| RandomForest    | 0.0216806238166419       | 0.09860414458809556      | 0.8838207219916316 |
| XGBoost         | 0.01771979285489721      | 0.08833842126303457      | 0.9050455024841315 |
```


In the presented results:

- **Best Parameters:** This section shows the hyperparameters that resulted in the best performance during the grid search cross-validation process for each model. These parameters were selected based on their ability to minimize the root mean squared error (RMSE) on the training data.

- **Best RMSE:** This indicates the RMSE achieved by each model using the best parameters selected during cross-validation. A lower RMSE indicates better predictive performance, as it means the model's predictions are closer to the actual values.

- **Model Evaluation:** The table provides the evaluation metrics for each model on the test set:
  - **Mean Squared Error (MSE):** This measures the average squared difference between the predicted and actual values. Lower MSE values indicate better model performance.
  - **Mean Absolute Error (MAE):** This measures the average absolute difference between the predicted and actual values. Lower MAE values indicate better model performance.
  - **R-squared (R²):** This represents the proportion of the variance in the dependent variable (sale prices) that is predictable from the independent variables (features) in the model. Higher R² values indicate a better fit of the model to the data.

Based on these metrics:

- The XGBoost model achieved the lowest RMSE, MSE, and MAE among the three models, indicating superior performance in terms of predictive accuracy.
- The XGBoost model also has the highest R-squared value, indicating that it explains more variance in the target variable compared to the other models.

Therefore, in this scenario, the XGBoost model is considered better because it has the lowest error metrics and the highest R-squared value, suggesting it provides the best balance of accuracy and explanatory power.

Exploring these variables for feature engineering can provide valuable insights into their distributions, relationships with the target variable, and potential transformations that could improve model performance. Here's what you can do with these variables:

1. **Missing Values**: Identify variables with a high percentage of missing values. You may need to handle these missing values appropriately, either by imputing them or dropping the corresponding columns.

2. **Categorical Variables**: Identify categorical variables among the listed features. For these variables, you can explore the cardinality (number of unique categories) and distribution of each category. Depending on the cardinality, you may need to apply techniques like one-hot encoding or ordinal encoding.

3. **Numerical Variables**: For numerical variables, you can examine their distributions, summary statistics (mean, median, standard deviation, etc.), and potential outliers. Outliers may need to be treated using techniques such as winsorization or robust scaling.

4. **Temporal Variables**: Variables like `GarageYrBlt` represent temporal information. You can explore their distributions over time, identify trends, and potentially derive new features such as age (e.g., current year - `GarageYrBlt`) or bins representing different time periods.

5. **Potential Interactions**: Investigate potential interactions or relationships between variables. For example, you might explore the relationship between `GarageCars` (number of cars in the garage) and `GarageArea` (garage size) to see if there's a consistent pattern.

6. **Correlation**: Examine the correlation between numerical variables to identify highly correlated pairs. High correlations may indicate multicollinearity, which can impact model performance.

By exploring these variables, you can gain a better understanding of your dataset and make informed decisions about preprocessing steps, feature engineering, and model selection.

In [57]:
var_explore = df[['Fence','Alley','MiscFeature','PoolQC','FireplaceQu','GarageCond','GarageQual','GarageFinish','GarageType','BsmtExposure','BsmtFinType2','BsmtFinType1','BsmtCond','BsmtQual','MasVnrType','Electrical','MSZoning','Utilities','Exterior1st','Exterior2nd','KitchenQual','Functional','SaleType','LotFrontage','GarageYrBlt','MasVnrArea','BsmtFullBath','BsmtHalfBath','GarageCars','GarageArea','TotalBsmtSF']]

display(HTML(create_scrollable_table(var_explore, 'var_explore', 'List of Variables to Explore for Feature Engineering')))

In [58]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import FunctionTransformer
import pandas as pd

# Define feature engineering functions 
def custom_features(df):
    df_out = df.copy()
    df_out['PropertyAge'] = df_out['YrSold'] - df_out['YearBuilt']
    df_out['TotalSF'] = df_out['TotalBsmtSF'] + df_out['1stFlrSF'] + df_out['2ndFlrSF']
    df_out['TotalBath'] = df_out['FullBath'] + 0.5 * df_out['HalfBath'] + df_out['BsmtFullBath'] + 0.5 * df['BsmtHalfBath']
    df_out['HasRemodeled'] = (df_out['YearRemodAdd'] != df_out['YearBuilt']).astype(object)
    df_out['Has2ndFloor'] = (df_out['2ndFlrSF'] > 0).astype(object)
    df_out['HasGarage'] = (df_out['GarageArea'] > 0).astype(object)
    df_out['YrSold_cat'] = df_out['YrSold'].astype(object)
    df_out['MoSold_cat'] = df_out['MoSold'].astype(object)
    df_out['YearBuilt_cat'] = df_out['YearBuilt'].astype(object)
    df_out['MSSubClass_cat'] = df_out['MSSubClass'].astype(object)
    
    return df_out

# Define a FunctionTransformer
feature_engineering_transformer = FunctionTransformer(custom_features)

# Identify new categorical and numerical columns
new_cols_categorical = ['HasRemodeled', 'Has2ndFloor', 'HasGarage']
new_cols_numeric = ['PropertyAge', 'TotalSF', 'TotalBath', 'YrSold_cat', 'MoSold_cat', 'YearBuilt_cat', 'MSSubClass_cat']

# Combine new columns with existing ones
categorical_columns = df.select_dtypes(include=['object', 'category']).columns.tolist() + new_cols_categorical
numerical_columns = df.select_dtypes(include=['int64', 'float64']).columns.tolist() + new_cols_numeric

# Remove target variable from numerical columns
numerical_columns.remove('SalePrice')

# Define transformers for numerical and categorical columns
numerical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

# Combine transformers using ColumnTransformer
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_columns),
        ('cat', categorical_transformer, categorical_columns)
    ],remainder = 'passthrough')

# Create a pipeline with the feature engineering, preprocessor, and PCA
pipeline_fe = Pipeline(steps=[
    ('fe', feature_engineering_transformer),
    ('preprocessor', preprocessor),
    ('pca', pca)
])

# Apply the pipeline to your dataset
X = df.drop('SalePrice', axis=1)
y = np.log(df['SalePrice'])
X_preprocessed_fe = pipeline_fe.fit_transform(X)

This code performs several preprocessing steps on a dataset, including feature engineering, data transformation, and dimensionality reduction using PCA (Principal Component Analysis). Here's a breakdown of what each part of the code does:

1. **Feature Engineering Functions**:
   - Defines a set of feature engineering functions (`custom_features`) that create new features based on existing ones. For example, it calculates the age of the property (`PropertyAge`), total square footage (`TotalSF`), total number of bathrooms (`TotalBath`), and binary indicators for whether the property has been remodeled (`HasRemodeled`), has a second floor (`Has2ndFloor`), or has a garage (`HasGarage`).
   
2. **Function Transformer**:
   - Uses the `FunctionTransformer` from scikit-learn to apply the feature engineering functions to the dataset. This transformer ensures that the custom feature engineering functions are compatible with scikit-learn's pipeline API.

3. **Identifying New Columns**:
   - Identifies the new categorical and numerical columns created by the feature engineering functions and combines them with the existing columns in the dataset.

4. **Column Transformer**:
   - Defines transformers for numerical and categorical columns using scikit-learn's `ColumnTransformer`. The numerical transformer performs imputation of missing values using the mean and scales the numerical features using standardization. The categorical transformer performs imputation using a constant value and one-hot encodes categorical features.

5. **Pipeline Creation**:
   - Combines the feature engineering transformer, the column transformer, and PCA into a single scikit-learn pipeline (`pipeline_fe`). This pipeline sequentially applies feature engineering, preprocessing, and PCA to the dataset.

6. **Pipeline Application**:
   - Applies the pipeline to the dataset (`X_preprocessed_fe`) to perform all the preprocessing steps, including feature engineering, data transformation, and dimensionality reduction using PCA.

Overall, this code efficiently preprocesses the dataset by applying feature engineering, data transformation, and dimensionality reduction techniques in a unified pipeline.

In [59]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train_fe, X_test_fe, y_train_fe, y_test_fe = train_test_split(X_preprocessed_fe, y, test_size=0.2, random_state=42)

# Define the models
models = {
    'LinearRegression': LinearRegression(),
    'RandomForest': RandomForestRegressor(random_state=42),
    'XGBoost': XGBRegressor(random_state=42)
}

# Define the hyperparameter grids for each model
param_grids = {
    'LinearRegression': {},
    'RandomForest': {
        'n_estimators': [100, 200, 500],
        'max_depth': [None, 10, 30],
        'min_samples_split': [2, 5, 10],
    },
    'XGBoost': {
        'n_estimators': [100, 200, 500],
        'learning_rate': [0.01, 0.1, 0.3],
        'max_depth': [3, 6, 10],
    }
}

# 3-fold cross-validation
cv = KFold(n_splits=3, shuffle=True, random_state=42)

# Train and tune the models
grids_fe = {}
for model_name, model in models.items():
    #print(f'Training and tuning {model_name}...')
    grids_fe[model_name] = GridSearchCV(estimator=model, param_grid=param_grids[model_name], cv=cv, scoring='neg_mean_squared_error', n_jobs=-1, verbose=2)
    grids_fe[model_name].fit(X_train_fe, y_train_fe)
    best_params = grids_fe[model_name].best_params_
    best_score = np.sqrt(-1 * grids_fe[model_name].best_score_)
    
    print(f'Best parameters for {model_name}: {best_params}')
    print(f'Best RMSE for {model_name}: {best_score}\n')

Fitting 3 folds for each of 1 candidates, totalling 3 fits
Best parameters for LinearRegression: {}
Best RMSE for LinearRegression: 0.16395567878468517

Fitting 3 folds for each of 27 candidates, totalling 81 fits
Best parameters for RandomForest: {'max_depth': None, 'min_samples_split': 5, 'n_estimators': 500}
Best RMSE for RandomForest: 0.15030523155118708

Fitting 3 folds for each of 27 candidates, totalling 81 fits
Best parameters for XGBoost: {'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 200}
Best RMSE for XGBoost: 0.13782341817422314



These results provide insights into the optimal hyperparameters selected for each model and their corresponding performance in terms of RMSE. It seems like XGBoost achieved the lowest RMSE among the three models, suggesting it might be the most effective model for this dataset.

In [60]:
# Provide the path to the test CSV file
test_file_path = r'C:\Users\nasrin\Desktop\DSproject\house-prices-advanced-regression-techniques\test.csv'

# Load the test dataset into a DataFrame
df_test = pd.read_csv(test_file_path)

df_test_preprocessed = pipeline_fe.transform(df_test)